### PR TITLE
[nomerge] Remove gyblets from FloatingPoint.swift

### PR DESCRIPTION
**Pull request by stephentyrone**
We only use .gyb features in two places in the FloatingPoint protocols, one of which we should be able to remove entirely: concrete inits from the builtin integer types, since those should be obsoleted by the BinaryInteger inits.

The other use is for random(in:), where it's only used to abstract over [Closed]Range. This mostly saves us some comment duplication, as the actually implementations are substantially different.

Also split BinaryFloatingPoint out into a new file, to make these files somewhat more manageable.

(Partially) Resolves [SR-5406](https://bugs.swift.org/browse/SR-5406), aka <rdar://problem/33199966>

### CODE REVIEW

**Comment by stephentyrone** [comment]
@swift-ci Please test

**Comment by stephentyrone** [comment]
@swift-ci Please smoke benchmark

**Comment by stephentyrone** [comment]
@swift-ci test source compatibility

File: stdlib/public/core/BinaryFloatingPoint.swift [hunk 1]
```diff
@@ -0,0 +1,677 @@
...
+    return (rounded, false)
+  }
+  
+  /// Creates a new value, rounded to the closest possible representation.
+  ///
+  /// If two representable values are equally close, the result is the value
+  /// with more trailing zeros in its significand bit pattern.
+  ///
+  /// - Parameter value: The integer to convert to a floating-point value.
+  @inlinable // FIXME(sil-serialize-all)
```
Comment (jrose-apple) [inline stdlib/public/core/BinaryFloatingPoint.swift]: You should probably upgrade this to `@_transparent` or at least `@inline(__always)`. This *definitely* has to get compiled down to a single instruction for common types, even in -Onone builds, right?

Comment (stephentyrone) [reply stdlib/public/core/BinaryFloatingPoint.swift]: Yeah, there's some follow-on work for the concrete types here, which is why this is [nomerge]. Just making sure that nothing's broken with the first step of this change.

Comment (stephentyrone) [reply stdlib/public/core/BinaryFloatingPoint.swift]: ... and actually there's a bunch of cases where we should be able to generate single-instruction conversions in generic contexts but fail to do so today even with optimization enabled, so we clearly need some more test coverage here.

File: stdlib/public/core/FloatingPoint.swift [hunk 2]
```diff
@@ -239,16 +230,6 @@ public protocol FloatingPoint : SignedNumeric, Strideable, Hashable
...
-% for src_ty in all_integer_types(word_bits):
-  /// Creates a new value, rounded to the closest possible representation.
-  ///
-  /// If two representable values are equally close, the result is the value
-  /// with more trailing zeros in its significand bit pattern.
-  ///
-  /// - Parameter value: The integer to convert to a floating-point value.
-  init(_ value: ${src_ty.stdlib_name})
-
-% end
```
Comment (natecook1000) [inline stdlib/public/core/FloatingPoint.swift]: Good riddance!

**Review by natecook1000 — APPROVED** [review]
🎉

File: stdlib/public/core/BinaryFloatingPoint.swift (hunk 1 above)
Comment (gottesmm) [reply stdlib/public/core/BinaryFloatingPoint.swift]: @jrose-apple I don't think we always inline at -Onone. If my memory is correct, this is something that I think would be good to put in.

File: stdlib/public/core/BinaryFloatingPoint.swift [hunk 3]
```diff
@@ -0,0 +1,677 @@
...
+    let (value_, exact) = Self._convert(from: value)
+    guard exact else { return nil }
+    self = value_
+  }
+  
+  @inlinable // FIXME(sil-serialize-all)
+  public // @testable
+  static func _convert<Source : BinaryFloatingPoint>(
+    from source: Source
+    ) -> (value: Self, exact: Bool) {
```
Comment (xwu) [inline stdlib/public/core/BinaryFloatingPoint.swift]: I think moving this to a different file has done something with the indentation here. I know Xcode tends to do this when copying and pasting.

File: stdlib/public/core/FloatingPoint.swift (hunk 2 above)
Comment (xwu) [reply stdlib/public/core/FloatingPoint.swift]: I'm *very* curious if, without this, the compiler can optimize the generic initializer down to a single instruction. If so, wow.

File: stdlib/public/core/BinaryFloatingPoint.swift [hunk 4]
```diff
@@ -0,0 +1,677 @@
...
+  /// - Parameter value: A floating-point value to be converted.
+  init(_ value: Float)
+  
+  /// Creates a new instance from the given value, rounded to the closest
+  /// possible representation.
+  ///
+  /// - Parameter value: A floating-point value to be converted.
+  init(_ value: Double)
+  
+  #if !os(Windows) && (arch(i386) || arch(x86_64))
```
Comment (xwu) [inline stdlib/public/core/BinaryFloatingPoint.swift]: Ditto here with the originally outdented compiler condition.

File: stdlib/public/core/FloatingPoint.swift (hunk 2 above)
Comment (stephentyrone) [reply stdlib/public/core/FloatingPoint.swift]: It very much cannot, but that will be addressed in part two =)

File: stdlib/public/core/BinaryFloatingPoint.swift (hunk 3 above)
Comment (stephentyrone) [reply stdlib/public/core/BinaryFloatingPoint.swift]: Yup. Command-option-shift-v preserves formatting though, FYI.

**Pull request closed without merging**